Support User Defines and Macro References AB Experiment

kuchungmsft opened the pull request:
Support A/B Compiler Arguments Traits

- Depends on https://github.com/microsoft/vscode-cpptools/pull/12979 to merge first.
- copilotcppMacroReferenceFilter: regex string to filter macro reference for telemetry.
- Telemetry related to user defines and macro references.
- Added new trait compilerUserDefines to note the relevant user defines to

kuchungmsft committed:
Support A/B Compiler Arguments Traits

- Depends on cpptools' update to provide ProjectContextResult.
- Added the following new traits
  - intellisense: compiler information disclaimer.
  - intellisenseBegin: to note the begin of IntelliSense information.
  - compilerArguments: a list of compiler command arguments that could affect Copilot generating completions.
  - directAsks: direct asking Copilot to do something instead of providing an argument.
  - intellisenseEnd: to note the end of IntelliSense information.
- A/B Experimental flags
  - copilotcppTraits: boolean flag to enable cpp traits
  - copilotcppExcludeTraits: string array to exclude individual trait, i.e., compilerArguments.
  - copilotcppMsvcCompilerArgumentFilter: regex string to match compiler arguments for GCC.
  - copilotcppClangCompilerArgumentFilter: regex string to match compiler arguments for Clang.
  - copilotcppGccCompilerArgumentFilter: regex string to match compiler arguments for MSVC.
  - copilotcppCompilerArgumentDirectAskMap: a stringify map string to map arguments to direct ask statements.

sean-mcmanus commented on Extension/src/LanguageServer/copilotProviders.ts:
```diff
@@ -59,6 +60,21 @@ export async function registerRelatedFilesProvider(): Promise<void> {
+                                if (configuration) {
+                                    const defines = configuration.defines?.join(', ') ?? undefined;
+                                    if (defines) {
+                                        traits.push({ name: "compilerDefines", value: defines, includeInPrompt: true, promptTextOverride: `This compiler defines contain: ${defines}.` });
```
Should "contain" be "are"? Same for compiler arguments below.

bobbrow replied:
Whatever you decide, this string probably needs to be localized, right?  Or is this a prompt for the AI (e.g. not a user-facing string)?

sean-mcmanus replied:
I assumed a prompt for the AI? In which case, it seems like it would be safer/better for the defines prompt to specifically mention that it's the "user-defined" defines so it doesn't think it includes the compiler defined ones.

sean-mcmanus replied:
`configuration.compilerArgs` and ` configuration.defines` only has configuration info for c_cpp_properties.json, which excludes configurations from configuration providers and compile_commands.json so it's not what you want to check. Colen said there's some function you can call on cpptools to get the info I think.

kuchungmsft replied:
@sean-mcmanus  I used 'contain` instead of `are` because the ones we provided here is subset. I have updated the implementation to get information from cpptools, please take another look, thanks.
@bobbrow this is a prompt for the AI, not a user-facing string, no need to localize. Thanks.

sean-mcmanus replied:
@kuchungmsft Why is it a subset? Which defines/args are missing?

sean-mcmanus replied:
It seems like maybe it should be a subset. Is there a max number of defines sent that I didn't notice?

kuchungmsft replied:
Since we'll be sending selective arguments once find out via A/Bs, will use "contain" instead of "are". Thanks.

Colengms commented on Extension/src/LanguageServer/copilotProviders.ts:
```diff
@@ -59,6 +60,21 @@ export async function registerRelatedFilesProvider(): Promise<void> {
 
+                            if (client instanceof DefaultClient) {
+                                const defaultClient: DefaultClient = client as DefaultClient;
+                                const configuration = defaultClient.configuration.CurrentConfiguration;
```
This is the 'base configuration'.  This may provide some information from some users' configurations, but a lot of users (perhaps the majority of users with non-trivial projects) do not configure the extension this way.  When configuring using a custom configuration provider (CMake Tools, Makefile tools), the 'base configuration' is effectively unused (and would be incorrect to refer to), as any file managed by those providers will get their configuration from them instead. (The base configuration is used as a fallback for files they don't handle).  Use of `compile_commands.json` to configure the extension is similar, as any file mentioned there will derived its configuration from that file.

I'd recommend extending the `cpptools_getChatContext` request to include any additional fields that might be useful, such as defines and compilerArgs.  That call will use the fully resolved configuration, so users of config providers and compile_commands.json would be supported.  However, that would be a fully resolved configuration, so could have tons of system defines. We would need to add some way for that to exclude defines retrieved from the compiler, as you probably wouldn't want those bloating the AI context.

sean-mcmanus replied:
@kuchungmsft I think you may be able to look at the compilation_args::original_defines data for the user-defines that exclude system defines.

kuchungmsft replied:
Thanks @Colengms  and @sean-mcmanus, I've updated the implementation, please take a look.

sean-mcmanus commented on Extension/test/scenarios/SingleRootProject/tests/lmTool.test.ts:
```diff
@@ -0,0 +1,206 @@
+        const result = await new moduleUnderTest.CppConfigurationLanguageModelTool().invoke(mockLanguageModelToolInvocationOptions, new vscode.CancellationTokenSource().token);
+
+        ok(result['text/plain'], 'result should contain a text/plain entry');
+        ok(result['text/plain'] === 'The user is working on a c++ project. The project uses language version C++20, compiles using the MSVC compiler, targets the Windows platform, and targets the x64 architecture.');
```
I think `c++` should be `C++`

sean-mcmanus replied:
Oh, but this is just a test? Can the source that uses `c++` be changed?

sean-mcmanus replied:
Copilot says:
![image](https://github.com/user-attachments/assets/21e25ccd-aedf-4a0e-8db8-537428ff4bc6)
so it seems better to use prompts it thinks are correct.

kuchungmsft replied:
Thanks, fixed.

Colengms commented on Extension/src/LanguageServer/client.ts:
```diff
@@ -539,6 +539,8 @@ export interface ChatContextResult {
     compiler: string;
     targetPlatform: string;
     targetArchitecture: string;
+    compilerArgs?: string[];
```
I'd just like to raise a flag here about including `compilerArgs` in the context. I don't think this is going to be helpful. In some cases, this will be full of 100's of arbitrary GCC/clang warning categories being selectively enabled and disabled.  My opinion is that this would be "distracting" in an AI context, giving it many, many things to focus on (and misinterpret as relevant) that would have little to do with the request.  I would relatively strongly advise against it.  Please pass along my feedback to whoever thinks this is a good idea.

sean-mcmanus replied:
Yeah, I think we should only include flags that we know are likely to affect the AI generated code or responses and not flags that just impact the compiler details (optimizations, warnings, etc.). And maybe each flag could be A/B tested separately.

sean-mcmanus replied:
Also are the "defines" going to be A/B tested separately so we can determine if that improves stuff or not?

sean-mcmanus replied:
And see my comment on the cpptools side.

kuchungmsft replied:
Thanks. I changed the PR to better support compiler arguments A/B experiments, so we'll know better which arguments to pass to Copilot. Which arguments to pass is controlled by experimental flags.

For user defines, I think it makes more sense to just pass the ones that are referenced in current document instead of all, need to check how to do that on cpptools side.

Colengms commented on Extension/src/LanguageServer/copilotProviders.ts:
```diff
@@ -52,6 +53,16 @@ export async function registerRelatedFilesProvider(): Promise<void> {
+
+                            const compilerUserDefines = chatContext.compilerUserDefines?.join(', ') ?? undefined;
+                            if (compilerUserDefines) {
+                                traits.push({ name: "compilerUserDefines", value: compilerUserDefines, includeInPrompt: true, promptTextOverride: `The compiler command line user defines contain: ${compilerUserDefines}.` });
```
Not added by you, but I'm just noticing this now.

"This project compiles using"
"This build targets"

We should avoid giving the AI false or misleading information. Those blurbs could easily be misinterpreted. We shouldn't tell the AI that "This project compiles...", as that could potentially incur some assumption that the code is valid as it "does compile", when it may not.

The C/C++ Extension does not include a build system. If there is an assumption that the user is "building" something, that "build" information is available, accurate and obviously applicable, as it is in VS, those are not safe assumption to make w/ the C/C++ Extension. It would probably be best not to mention "build" in the context, so as not to suggest to the AI that the user is building the code or may need assistance related to the actual build process. 

`targetPlatform` and `targetArchitecture` refer to the platform and architecture IntelliSense in configured for, not necessarily a build system. I'd suggest using something technically correct without overreaching, such as "The C++ language standard is currently set to:", and "This code may target:", or something else not potentially incorrect.

kuchungmsft replied:
Thanks. Will start A/B with the statements that are already checked in, and future A/Bs to find better statements, if any.

benmcmorran commented on Extension/src/LanguageServer/lmTool.ts:
```diff
@@ -67,20 +67,26 @@ export class CppConfigurationLanguageModelTool implements vscode.LanguageModelTo
+                telemetryProperties["compilerArgs"] = chatContext.compilerArgs.join(' ');
+            }
+            if (chatContext.compilerUserDefines) {
+                telemetryProperties["compilerUserDefines"] = chatContext.compilerUserDefines.join(', ');
```
`compilerUserDefines` and `compilerArgs` could both contain PII. To avoid privacy issues, can we only emit telemetry for values that come from an allow list of well-known defines or arguments?

benmcmorran replied:
Discussed offline with @bobbrow. Let's drop this `compilerUserDefines` and `compilerArgs` telemetry for now. We can add it back later in a more targeted way if there are specific flags we need data on.

kuchungmsft replied:
Updated telemetry to only log the number of arguments and number of arguments being sent to copilot. Thanks.

bobbrow commented on Extension/src/LanguageServer/lmTool.ts:
```diff
@@ -44,6 +47,85 @@ const knownValues: { [Property in keyof ChatContextResult]?: { [id: string]: str
+}
+
+// To be updated after A/B experiments.
+const defaultCompilerArgumentFilters: { [id: string]: RegExp | undefined } = {
```
nit: `undefined` means the key doesn't actually exist in the object.  This might be defined like this instead:
```ts
const defaultCompilerArgumentFilters: { [id: string]: RegExp } = {};
```

bobbrow replied:
(though from what I can tell, this object never adds defaults from any of the compilers and will remain empty)

kuchungmsft replied:
Updated. Will update defaults after A/B experiemnts, we'll know better on what arguments we should provide to help completions.

sean-mcmanus commented on Extension/src/LanguageServer/lmTool.ts:
```diff
@@ -44,6 +47,84 @@ const knownValues: { [Property in keyof ChatContextResult]?: { [id: string]: str
+    return {
+        language: chatContext.language,
+        standardVersion: chatContext.standardVersion,
+        compiler: chatContext.compiler,
```
I'm worried that this will tend to generate code that only compiles for the 1 compiler/platform/architecture, and the A/B testing may be insufficient to detect that code is generated isn't compatible, because users won't realize the code doesn't compile with other targets until later on when there's a compile failure since IntelliSense won't give errors since it's only configured for the 1 target. The only reason IntelliSense only has 1 target configured is because supporting IntelliSense for multiple targets would be too expensive for performance.

According to Copilot: 
![image](https://github.com/user-attachments/assets/753059db-5967-41fd-ad22-ed55587db9bd)

sean-mcmanus replied:
Here's a specific case where Copilot suggested code that only compiles with gcc/mingw, but when I switch my IntelliSense config to msvc/cl.exe it fails.

Copilot
<img width="548" alt="image" src="https://github.com/user-attachments/assets/c51a39fb-293d-40ff-a650-25392867bf92">

gcc IntelliSense config:
![image](https://github.com/user-attachments/assets/17ace44a-3850-40b3-90b6-4685c52d226a)

msvc IntelliSense config:
<img width="526" alt="image" src="https://github.com/user-attachments/assets/069ec12b-2757-47e3-8dba-e07d3c7deff8">

kuchungmsft replied:
@sean-mcmanus can you show me the code snippet before/after the Copilot suggestion?

kuchungmsft replied:
Sean, I was going to link these cross-plat related comments to the bug that Ben created this morning, it looks like you already added the comments. I'll go ahead and resolve this thread as it's out of scope for this PR. Thanks.

sean-mcmanus commented on Extension/src/LanguageServer/lmTool.ts:
```diff
@@ -72,12 +153,7 @@ export class CppConfigurationLanguageModelTool implements vscode.LanguageModelTo
-            }
+            formatChatContext(chatContext);
 
             return `The user is working on a ${chatContext.language} project. The project uses language version ${chatContext.standardVersion}, compiles using the ${chatContext.compiler} compiler, targets the ${chatContext.targetPlatform} platform, and targets the ${chatContext.targetArchitecture} architecture.`;
```
...as mentioned in my other comment, sending `The project uses language version ${chatContext.standardVersion}, compiles using the ${chatContext.compiler} compiler, targets the ${chatContext.targetPlatform} platform, and targets the ${chatContext.targetArchitecture} architecture.`` could potentially mislead Copilot into believing the project is not cross-plat or intended to compile for multiple compilers, platforms, architectures.

sean-mcmanus replied:
![image](https://github.com/user-attachments/assets/c1cbf79e-2c79-45ad-95b9-68f1da142108)

sean-mcmanus replied:
![image](https://github.com/user-attachments/assets/a882f47e-3b1f-4dd8-a16c-5cb3479dfa42)

sean-mcmanus replied:
It seems like we should only tell Copilot what a project is intended to target if we actually know that info. The IntelliSense context just reports what IntelliSense is currently configured for in regards to real-time language service features. Users might have multiple IntelliSense configurations for multiple targets (with IntelliSense being configured for only 1 at a time for performance reasons).

kuchungmsft replied:
Thanks Sean. This statement is only provided to support A/B experiments of Copilot Chat experience (not completions). When users type '#cpp' in the chat window, we'll send this project context to Copilot to help generating better response, if any. This is out of scope as my PR is focusing on supporting A/B to check if sending any of compiler arguments would help Copilot generating better completions. Ben already responded your email, and created a bug to track the work to make Copilot to be more aware of cross-plat context.

Colengms commented on Extension/src/LanguageServer/lmTool.ts:
```diff
@@ -44,6 +47,84 @@ const knownValues: { [Property in keyof ChatContextResult]?: { [id: string]: str
+// To be updated after A/B experiments, math nothing for now.
+const defaultCompilerArgumentFilters: { [id: string]: RegExp } = {
+    MSVC: matchNothingRegex,
+    Clang: matchNothingRegex,
```
One thing to consider is that on macOS the default compiler is clang, and `/usr/bin/gcc` / `/usr/bin/g++` are actually symbolic links to clang, not gcc.  Performing a regex match on the compiler path may not be a reliable way to determine what compiler is being referred to. On the native side, we 'query' non-CL compilers (gcc or clang), and will use system defines to determine whether the compiler is clang or gcc.  (There could also be other details that could be determined based on detected system defines, that may potentially be worth surfacing to AI in the future).  I'd suggest adding plumbing on the native side to pass back the 'compiler type' (with the rest of the TU context) instead of relying on pattern matching against compiler paths.

kuchungmsft replied:
These regex are for matching the compiler arguments to be sent to the Copilot, these are not for detecting the compiler types.

Colengms replied:
I see.  FYI, the source of "compiler" here appears to be coming from the `IntelliSenseMode` on the native side, only.  It's not quite correct to directly associate that with compiler arguments.  For instance, clang has a compiler, `clang-cl.exe`, which is clang, but accepts MSVC command line arguments (as well as some clang arguments not accepted by MSVC).  The user will get a (i.e.) `windows-clang-x64` intelliSense-mode, despite all of their arguments being CL arguments.

kuchungmsft replied:
Thanks, Colen, I didn't know that, I think we can configure the regex to match the arguments from both scenarios for Clang so that we can cover `windows-clang-XXX` as well. So, I think we are good.

kuchungmsft commented on Extension/src/LanguageServer/lmTool.ts:
```diff
@@ -44,6 +48,102 @@ const knownValues: { [Property in keyof ChatContextResult]?: { [id: string]: str
+        telemetryProperties["compilerArgumentCount"] = projectContext.result.fileContext.compilerArgs.length.toString();
+        telemetryProperties["filteredCompilerArgumentCount"] = filteredcompilerArguments.length.toString();
+        telemetryProperties["compilerUserDefinesRelavantCount"] = projectContext.result.fileContext.compilerUserDefinesRelavant.length.toString();
+        telemetryProperties["targetArcompilerOriginalUserDefineCounthitecture"] = projectContext.result.fileContext.compilerOriginalUserDefineCount.toString();
```
Typo: compilerOriginalUserDefineCount

kuchungmsft commented on Extension/src/LanguageServer/lmTool.ts:
```diff
@@ -44,6 +48,102 @@ const knownValues: { [Property in keyof ChatContextResult]?: { [id: string]: str
+        telemetryProperties["filteredCompilerArgumentCount"] = filteredcompilerArguments.length.toString();
+        telemetryProperties["compilerUserDefinesRelavantCount"] = projectContext.result.fileContext.compilerUserDefinesRelavant.length.toString();
+        telemetryProperties["targetArcompilerOriginalUserDefineCounthitecture"] = projectContext.result.fileContext.compilerOriginalUserDefineCount.toString();
+        telemetryProperties["targetArchmacroReferenceCountitecture"] = projectContext.result.fileContext.macroReferenceCount.toString();
```
Typo: macroReferenceCount

sean-mcmanus commented on Extension/src/LanguageServer/client.ts:
```diff
@@ -541,6 +541,22 @@ export interface ChatContextResult {
 
+export interface FileContextResult {
+    compilerArgs: string[];
+    compilerUserDefinesRelavant: string[];
```
Relavant->Relevant

kuchungmsft replied:
Fixed, thanks.

kuchungmsft committed:
Support User Defines and Macro References AB Experiment

- Depends on https://github.com/microsoft/vscode-cpptools/pull/12979 to merge first.
- copilotcppMacroReferenceFilter: regex string to filter macro reference for telemetry.
- Telemetry related to user defines and macro references.
- Added new trait compilerUserDefines to note the relevant user defines to the editing file.
